File: rust/src/bin/p0029.rs
// Project Euler: Problem 29

// Note that use `for` loops for clarity in this solution

euler::run_solver!(29);

fn solve() -> String {
    compute(100).to_string()
}

fn compute(upper: usize) -> u64 {
    use euler::math;

    debug_assert!(upper >= 2);

    let dupctr_tbl = make_dupctr_tbl(upper);
    let base_limit = upper.isqrt();
    let mut skip_flag = vec![false; base_limit + 1];
    let mut ans = (upper - 1).pow(2) as u64;

    for b in 2..=base_limit {
        if skip_flag[b] {
            continue;
        }
        for (e, &elm) in dupctr_tbl
            .iter()
            .enumerate()
            .take(math::get_max_exp(upper as u64, b as u64) as usize + 1)
            .skip(2)
        {
            ans -= elm;
            let tmp = b.pow(e as u32);
            if tmp <= base_limit {
                skip_flag[tmp] = true;
            }
        }
    }

    ans
}

fn make_dupctr_tbl(upper: usize) -> Vec<u64> {
    use euler::math;

    let max_exp = math::get_max_exp(upper as u64, 2) as usize;
    let mut dupctr_tbl = vec![0_u64; max_exp + 1];

    for (x, elm) in dupctr_tbl.iter_mut().enumerate().take(max_exp + 1).skip(2) {
        let mut dups = vec![0_u64; upper + 1];
        for y in 1..x {
            let k = math::lcm(x as u64, y as u64) as usize / x;
            (k.max(2)..=(upper * y / x))
                .step_by(k)
                .for_each(|idx| dups[idx] = 1);
        }
        *elm = dups.iter().sum();
    }

    dupctr_tbl
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0029_upper_5() {
        assert_eq!(compute(5), 15);
    }

    #[test]
    fn p0029_upper_100() {
        assert_eq!(compute(100), 9183);
    }
}

File: ocaml/src/p0029.ml
(* Project Euler: Problem 29 *)

open Core

let make_dupctr_tbl upper =
  let max_exp = Euler.Math.get_max_exp upper ~base:2 in
  let tbl = Array.create 0 ~len:(max_exp + 1) in
  Sequence.(
    range 2 max_exp ~stop:`inclusive
    |> iter ~f:(fun x ->
      let dups = Array.create 0 ~len:(upper + 1) in
      range 1 x ~stop:`exclusive
      |> iter ~f:(fun y ->
        let k = Euler.Math.lcm x y / x in
        range (max k 2) (upper * y / x) ~stop:`inclusive ~stride:k
        |> iter ~f:(fun idx -> dups.(idx) <- 1));
      tbl.(x) <- Array.to_list dups |> List.reduce_exn ~f:( + )));
  tbl
;;

let compute upper =
  let dup_ctr_tbl = make_dupctr_tbl upper in
  let base_limit = Euler.Math.isqrt upper in
  let skip_flag = Array.create false ~len:(base_limit + 1) in
  let ans = ref ((upper - 1) * (upper - 1)) in
  Sequence.(
    range 2 base_limit ~stop:`inclusive
    |> iter ~f:(fun b ->
      if not skip_flag.(b)
      then
        range 0 (Euler.Math.get_max_exp upper ~base:b) ~stop:`inclusive
        |> Fun.flip drop 2
        |> iter ~f:(fun e ->
          ans := !ans - dup_ctr_tbl.(e);
          let tmp = Int.pow b e in
          if tmp <= base_limit then skip_flag.(tmp) <- true)));
  !ans
;;

let solve () = compute 100 |> Int.to_string

(* Test *)

let%test_unit "100" = [%test_eq: int] (compute 100) 9183
